File: Projeto/airqualityAPI/src/main/java/ptua/airqualityAPI/Services/AirQualityService.java
package ptua.airqualityAPI.Services;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import ptua.airqualityAPI.API.AQICN_API;
import ptua.airqualityAPI.Cache.Cache;
import ptua.airqualityAPI.Models.AirMetricsData;

import java.util.logging.Level;
import java.util.logging.Logger;

@Service
public class AirQualityService {

    @Autowired
    private AQICN_API aqicn_api;

    private final Cache cache = new Cache(10 * 60L);  //10 minutes of TTL

    private static final Logger logger = Logger.getLogger(AirQualityService.class.getName()); //Creation of logger

    public AirMetricsData getAirMetricsCity(String city) {
        logger.log(Level.INFO, "Trying to get data from cache");
        AirMetricsData data = cache.getRequest(city);

        if ( data == null ){
            logger.log(Level.INFO, "Cache didn't have the data. Requesting from the external API");
            data = this.aqicn_api.getAirMetris_City(city);
            cache.storeRequest(city, data);
        } else {
            logger.log(Level.INFO, "Data retrieved from the Cache");
        }

        return data;
    }

    public AirMetricsData getAirMetricsLatLong(Double latitude, Double longitude) {
        logger.log(Level.INFO, "Trying to get data from cache");
        AirMetricsData data = cache.getRequest(latitude,longitude);

        if ( data == null ){
            logger.log(Level.INFO, "Cache didn't have the data. Requesting from the external API");
            data = this.aqicn_api.getAirMetris_LatANDLong(latitude,longitude);
            cache.storeRequest(latitude, longitude, data);
        } else {
            logger.log(Level.INFO, "Data retrieved from the Cache");
        }

        return data;
    }

    public Cache getCacheStatistics(){
        return cache;
    }
}

File: Projeto/airquality_app/src/Services/AirQualityService.js
class AirQualityService {
    async getAirMetricsByCity(BASE_URL, city) {
        var url = BASE_URL + "getAirMetricsCity?city=" + city;


        var result = await fetch(url);

        return result.json();
    }

    async getAirMetricsByLatLong(BASE_URL, latitude, longitude) {
        var url = BASE_URL + "getAirMetricsLatitudeLongitude?latitude=" + latitude + "&longitude=" + longitude;

        var result = await fetch(url);

        return result.json();
    }

    async getCacheStatistics(BASE_URL) {
        var url = BASE_URL + "getCacheStatistics";

        var result = await fetch(url);

        return await result.json();
    }

}

export default new AirQualityService();
